jablib/src/main/java/org/jabref/logic/externalfiles/LinkedFileTransferHelper.java

```diff
@@ -0,0 +1,214 @@
+                }
+
+                // In case the file wasn't found at the source or the target - just re-use the existing link
+                Optional<Path> sourcePathOpt = linkedFile.findIn(sourceContext, filePreferences);
+                Optional<Path> targetPrimaryOpt = getPrimaryPath(targetContext, filePreferences);
+
+                if (sourcePathOpt.isEmpty() || targetPrimaryOpt.isEmpty()) {
+                    linkedFiles.add(linkedFile);
+                    continue;
+                }
+
+                Path relative;
+                if (sourcePathOpt.get().startsWith(targetPrimaryOpt.get())) {
+                    relative = targetPrimaryOpt.get().relativize(sourcePathOpt.get());
+                } else {
+                    relative = Path.of("..").resolve(sourcePathOpt.get().getFileName());
+                }
+
+                if (isReachableFromPrimaryDirectory(relative)) {
+                    // [impl->req~logic.externalfiles.file-transfer.reachable-no-copy~1]
+                    entryChanged = isPathAdjusted(linkedFile, relative, linkedFiles, entryChanged);
+                } else {
+                    entryChanged = isFileCopied(context, linkedFile, linkedFiles, entryChanged);
+                }
+            }
+            if (entryChanged) {
+                entry.setFiles(linkedFiles);
+                modifiedEntries.add(entry);
+            }
+        }
+        return modifiedEntries;
+    }
+
+    /**
+     * Gets the primary directory path for the given context.
+     * This is a utility method extracted from the original implementation.
+     *
+     * @param context The database context
+     * @param filePreferences File preferences for the context
+     * @return Optional containing the primary directory path, or empty if none found
+     */
+    public static Optional<Path> getPrimaryPath(BibDatabaseContext context, FilePreferences filePreferences) {
+        List<Path> directories = context.getFileDirectories(filePreferences);
+        if (directories.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(directories.getFirst());
+    }
+
+    /**
+     * Determines if the given relative path is reachable from the primary directory.
+     * A path is considered reachable if it does not start with ".." (i.e., does not traverse up the directory tree)
+     * and is not absolute.
+     *
+     * @param relativePath the path to check, relative to the primary directory
+     * @return true if the path is reachable from the primary directory, false otherwise
+     */
+    public static boolean isReachableFromPrimaryDirectory(Path relativePath) {
+        try {
+            return !relativePath.startsWith("..") && !relativePath.isAbsolute();
+        } catch (IllegalArgumentException e) {
+            return false;
+        }
+    }
+
+    private static boolean isPathAdjusted(LinkedFile linkedFile, Path relative, List<LinkedFile> linkedFiles, boolean entryChanged) {
```

Comment:
The method name is not good - it is a property of something - and reading (!) a property should not make changes to something.

Maybe, `adjustPathAndLinkFile`, because this method also adds the file to the list of LinkedFiles.